src/include/pka_cpu.h

```diff
@@ -80,14 +86,191 @@ static inline uint64_t pka_early_cpu_speed(void)
 
+struct dmi_header {
+    uint16_t handle;
+    uint8_t  type;
+    uint8_t  length;
+    uint8_t *data;
+};
+
+/// Global variable holding the cpu frequency.
+uint64_t cpu_f_hz;
+
+static void *pka_cpu_file_read(size_t *len, const char *fname)
+{
+    size_t   rd_cnt = 0;
+    ssize_t  rd_st  = 1;
+    struct   stat sbuf;
+    uint8_t *f;
+    int      fd;
+
+    if ((fd = open(fname, O_RDONLY)) == -1) {
+        if (errno)
+            printf("\n Opening file failed with error:%d.\n", errno);
+        return NULL;
+    }
+
+    if (fstat(fd, &sbuf) == 0) {
+        *len = sbuf.st_size;
+    }
+
+    if ((f = malloc(*len)) == NULL) {
+        printf("\n Memory alloc failed when reading file.\n");
+        goto out;
+    }
+
+    while (rd_cnt != *len && rd_st != 0) {
+        rd_st = read(fd, f+rd_cnt, *len - rd_cnt);
+        if (rd_st == -1) {
+            printf("\n Reading file failed.\n");
+            goto err;
+        }
+        else
+            rd_cnt += rd_st;
+    }
+
+    if (rd_cnt != *len) {
+        printf("\n EOF reached before expected length.\n");
+        goto err;
+    }
+    else
+        goto out;
+
+err:
+    free(f);
+    f = NULL;
+out:
+    if (close(fd) == -1)
+        printf("\n Closing the file failed.\n");
+
+    return f;
+}
+
+static int pka_cpu_checksum(const uint8_t *buf, size_t len)
+{
+    uint8_t sum = 0;
+    size_t  d;
+
+    for (d = 0; d < len; d++)
+         sum += buf[d];
+
+    return sum == 0;
+}
+
+static uint16_t pka_cpu_read_dmi_table(uint32_t len, uint16_t num,
+                               const char *memdev)
+{
+    uint8_t *buf, *data, *next;
+    struct   dmi_header h;
+    size_t   size;
+    int      i = 0;
+
+    if ((buf = pka_cpu_file_read(&size, memdev)) == NULL)
+        return 0;
+
+    len  = size;
+    data = buf;
+
+    while ((i < num || !num)
+        && data + 4 <= buf + len) {
+
+        /// Copy the header information
+        h.type   = data[0];
+        h.length = data[1];
+        h.handle = WORD(data + 2);
+        h.data   = data;
+
+        if (h.length < 4) {
+            printf("\n Invalid entry length, DMI table is broken.\n");
+            break;
+        }
+        i++;
+
+        if (h.type == 127)
+            break;
+
+        next = data + h.length;
+        while ((unsigned long)(next - buf + 1) < len
+            && (next[0] != 0 || next[1] != 0))
+                next++;
+
+        next += 2;
+
+        if ((unsigned long)(next - buf) > len) {
+            printf("\n Structure is truncated at %dth structure.\n", i);
+            break;
+        }
+
+        /// Processor information
+        if (h.type == 4) {
+            if (h.length < 0x1A) {
+                printf("\n Incomplete processor information.\n");
+                return 0;
+            }
+            /// Frequency is at 0x16 offset
+            return WORD(h.data+0x16);
+        }
+
+        data = next;
+    }
+    return 0;
+}
+
+static uint64_t pka_cpu_get_frequency()

```

Comment:
Could you please rename the function. For example: pka_cpu_get_freq_from_smbios()